EIPS/eip-9042.md

```diff
@@ -0,0 +1,94 @@
+---
+eip: 9042
+title: Create delegate
+description: Add a new EVM instruction allowing contracts to create clones using EIP-7702 delegation designations
+author: Hadrien Croubois (@amxx), Danno Ferrin (@shemnon)
+discussions-to: https://ethereum-magicians.org/t/eip-9042-create-delegate/21763
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-05-07
+requires: 7692, 7702
+---
+
+## Abstract
+
+Introduce a new EVM instruction to [EOF1](./eip-7692.md) that allows smart contracts to create (and update) delegation accounts that match [EIP-7702](./eip-7702.md)'s design. These accounts can be used similarly to [ERC-1167](./eip-1167.md) clones, with significant advantages.
+
+## Motivation
+
+Many on-chain applications involve creating multiple instances of the same code at different locations. These applications often rely on clones, or proxies, to reduce deployment costs.
+
+Clones, such as the one described in ERC-1167 are minimal pieces of code that contain the target address directly in the code. That makes them extremely light but prevents any form of reconfiguration (upgradability).
+
+Upgradeable proxies differ from clones in that they read the implementation's address from storage. This makes them more versatile but also more expensive to use.
+
+In both cases delegating the received calls to an implementation using EVM code comes with some downsides:
+
+- Calldata must be copied to memory before performing the delegate call
+- Clones and proxy written in EOF do not support delegation to an implementation written in legacy EVM code, and are thus limited or possibly dangerous. This encourages the continued use of legacy EVM code.
+
+EIP-7702 introduces a new type of object that has the expected behavior: designator. These objects are designed to be instantiated at the address of EOA's, but the same behavior could be re-used to implement clones at the protocol level. Using designators for this use-case provides upgradeability without the need for storage lookups if the contract calling the `CREATE_DELEGATE` allows it. It also removes any issue related to code version incompatibilities.
+
+## Specification
+
+A new instruction (`CREATE_DELEGATE`) is added to EOF1 at `0xf6`.
+
+### Behavior
+
+Executing this instruction does the following:
+
+1. deduct `EMPTY_ACCOUNT_COST` gas
+2. halt if the current frame is in `static-mode`
+3. pop `salt`, `target` from the operand stack
+4. calculate `location` as `keccak256(0xfe10000 ++ address ++ salt)[12:]`
+5. halt if the code at `location` is not empty and does not start with `0xEF0100` (no empty and not a designator)
+6. add `EMPTY_ACCOUNT_COST - BASE_COST` gas to the global refund counter if `location` exists in the trie.
+7. set the code of `authority` to be `0xef0100 || target`, matching the delegation process defined in EIP-7702.
```

Comment:
Would you consider `0xef0101` to differentiate EOA delegates from CREATE_DELEGAGES?  If this option is rejected can you add to the rationale why this would not be adopted?

Reply:
I don't have strong opinions either way.

Do you think EOA delegations and delegations created using this opcode should be distinguishable?

Reply:
For security auditing it may be nice to identify ones that can be upgrade via a known key and ones that can only be upgraded by contract.  But the ones that are contract upgradable would also need to know the contract that can do the upgrading, so they are already going to a block explorer at that point to get the source contract.  Explorers should be able to track the difference if they care.  I think when it goes for community review this is something we should ask if they care to easily distinguish between them in audits and forensics.

I think we can dismiss the concern of address collisions as too unlikely for now. i.e. a contract address where a suitable PrivKey is discovered.

Reply:
Yes, collisions should not be an issue here, at least not more than with create2